### PR TITLE
Prefer `itertools.pairwise()` over `zip()` for successive pairs (`RUF007`)

### LINKED ISSUE #3297: Use itertools.pairwise instead of zipping a sequence against itself
Tonight I discovered [`itertools.pairwise`][0] (new in 3.10) and replaced 6 lines like this in a ~10,000 line codebase:

```python
# Before:
for first, second in zip(my_sequence, my_sequence[1:], strict=False):
    ...

# After:
for first, second in pairwise(my_sequence):
    ...
```

I have no idea how easy this would be. But if it is easy it might make a nice rule!

And in addition to being less verbose, `pairwise` is also faster on my machine.

```python
import timeit
from itertools import pairwise

short = [0] * 10
medium = [0] * 1_000
long = [0] * 1_000_000

print(timeit.timeit("list(pairwise(short))", globals=globals()))
print(timeit.timeit("list(zip(short, short[1:], strict=False))", globals=globals()))
# 0.5838822979985707
# 1.1156816439997783

print(timeit.timeit("list(pairwise(medium))", globals=globals(), number=10_000))
print(timeit.timeit("list(zip(medium, medium[1:], strict=False))", globals=globals(), number=10_000))
# 0.36417481299940846
# 0.4547197800002323

print(timeit.timeit("list(pairwise(long))", globals=globals(), number=10))
print(timeit.timeit("list(zip(long, long[1:], strict=False))", globals=globals(), number=10))
# 1.0911181840001518
# 1.499006936999649
```

Another common idiom for this uses indexes. But my gut tells me linting this would be much more difficult and error prone.

```python
for i in range(len(my_sequence) - 2):
    first, second = my_sequence[i], my_sequence[i + 1]
    ...
```

[0]: https://docs.python.org/3/library/itertools.html#itertools.pairwise

### PR DESCRIPTION
**Pull request by evanrittenhouse**
Fixes #3297. 

We ensure that the two arguments to `zip()` match and that the lower bound of the slice is 1, if the second argument is a slice. Otherwise, we wouldn't be able to guarantee that the request being made is for successive pairs.

### CODE REVIEW

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let PairwiseOverZipped { expr, .. } = self;
+        format!("{:?}", expr)
+    }
+}
+
+pub fn pairwise_over_zipped(checker: &mut Checker, func: &Expr, args: &[Expr]) {
+    if let ExprKind::Name { id, .. } = &func.node {
+        if id == "zip" {
```
Comment (andersk) [inline crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs:26]: Also check `checker.ctx.is_builtin("zip")`.

**Review by andersk — COMMENTED** [review]
[`itertools.pairwise`](https://docs.python.org/3/library/itertools.html#itertools.pairwise) is new in Python 3.10, so this rule should be restricted to `target_version >= PyVersion::Py310`.

Should we also check for `zip(a[:-1], a[1:])`?

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs (hunk 1 above)
Comment (evanrittenhouse) [reply crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs:26]: Makes sense, I can implement these changes tomorrow

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    fn message(&self) -> String {
+        format!("Prefer `itertools.pairwise()` to `zip()` if trying to fetch an iterable's successive overlapping pairs.")
+    }
+}
+
+#[derive(Debug)]
+struct SliceInfo {
+    arg_name: String,
+    slice_start: i32,
+    slice_end: i32,
```
Comment (MichaReiser) [inline crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs:25]: I'm not as familiar with python but should these be `i64` to support a broader range of valid int indices?

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    Some(SliceInfo::new(arg_id.to_string(), lower_bound, upper_bound))
+}
+
+pub fn pairwise_over_zipped(checker: &mut Checker, func: &Expr, args: &[Expr]) {
+    if let ExprKind::Name { id, .. } = &func.node {
+        // Ensure that the checker settings are valid for the rule to apply
+        let valid_checker =
+            checker.ctx.is_builtin(id) && checker.settings.target_version >= PythonVersion::Py310;
+
+        if valid_checker && id == "zip" && args.len() > 1 {
```
Comment (MichaReiser) [inline crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs:83]: Checking if `ctx.is_builtin` is probably the most expensive operation (not very, but still):

We can help the compiler to generate more optimized code by changing the order of checks:

1. Test if the python version is valid 
2. Test if it only has one argument
3. Test if the id is `zip`
4. Only then, test if it is the built-in zip

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            // Second arg can only be a subscript
+            let ExprKind::Subscript { .. } = &args[1].node else {
+                return;
+            };
+            let second_arg_info = get_slice_info(&args[1], checker.stylist).unwrap();
+
+            let first_arg_info = first_arg_info_opt.unwrap();
+            let args_are_successive = (first_arg_info.slice_start == 0
+                || first_arg_info.slice_end == -1)
+                && second_arg_info.slice_start == 1;
```
Comment (MichaReiser) [inline crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs:105]: Let's avoid the `first_arg_info_opt.unwrap()`
```suggestion
            // If it's not one of those, return
            let Some(first_arg_info) = first_arg_info_opt else  { return; };

            // Second arg can only be a subscript
            let ExprKind::Subscript { .. } = &args[1].node else {
                return;
            };
            let second_arg_info = get_slice_info(&args[1], checker.stylist).unwrap();

            let args_are_successive = (first_arg_info.slice_start == 0
                || first_arg_info.slice_end == -1)
                && second_arg_info.slice_start == 1;
```

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            // If it's not one of those, return
+            if first_arg_info_opt.is_none() {
+                return;
+            }
+
+            // Second arg can only be a subscript
+            let ExprKind::Subscript { .. } = &args[1].node else {
+                return;
+            };
+            let second_arg_info = get_slice_info(&args[1], checker.stylist).unwrap();
```
Comment (MichaReiser) [inline crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs:100]: Can we add `// SAFETY` comment describing why it's safe to call `unwrap` here (or why it is guaranteed that `get_slice_info` always returns `Some` for `args[1])?

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    let mut upper_bound = 0;
+    if let ExprKind::Slice { lower, upper, .. } = &slice.node {
+        if lower.is_some() {
+            if let ExprKind::Constant {
+                value: lower_value, ..
+            } = &lower.as_ref().unwrap().node
+            {
+                lower_bound = unparse_constant(lower_value, stylist)
+                    .parse::<i32>()
+                    .unwrap_or(0);
```
Comment (andersk) [inline crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs:58]: We can’t just default unrecognized things to `0`. `a[0:-1]` and `a[2147483648:-1]` are quite different but this function treats them identically.

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs (hunk 5 above)
Comment (evanrittenhouse) [reply crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs]: Added a `return` if `get_slice_info` is `None` for `args[1]` which should remove the potential for panic

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs (hunk 3 above)
Comment (evanrittenhouse) [reply crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs]: Done

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs (hunk 2 above)
Comment (evanrittenhouse) [reply crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs]: Done

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    let ExprKind::Subscript { value, slice, .. } = &expr.node else {
+        return None;
+    };
+
+    let ExprKind::Name { id: arg_id, .. } = &value.node else {
+        return None;
+    };
+
+    let mut lower_bound = Ok(0i64);
+    let mut upper_bound = Ok(0i64);
```
Comment (andersk) [inline crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs:55]: We can’t default *these* unrecognized things to 0 either: `a[x:-1]` is different from `a[0:-1]`.

So is `a[0:-1:2]`, by the way.

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    arg_name: String,
+    slice_start: Result<i64, ParseIntError>,
+    slice_end: Result<i64, ParseIntError>,
+}
+
+impl SliceInfo {
+    pub fn new(
+        arg_name: String,
+        slice_start: Result<i64, ParseIntError>,
+        slice_end: Result<i64, ParseIntError>,
```
Comment (MichaReiser) [inline crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs:34]: It seems that we only care about whether parsing was successful or not, but not what the actual parsing error was. That means we could use `Option<i64>` here (you can use `result.ok()` to get an option) and safe a few bytes of memory.

**Review by MichaReiser — APPROVED** [review]

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs (hunk 7 above)
Comment (MichaReiser) [reply crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs:55]: @andersk brought up a few good edge cases. Could you add those to the test suite?

Comment (evanrittenhouse) [reply crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs:55]: Yup, will do

File: crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs (hunk 8 above)
Comment (evanrittenhouse) [reply crates/ruff/src/rules/ruff/rules/pairwise_over_zipped.rs:34]: Yeah, that's the path I'm going down in the latest commit. Seems like it'll help with the default issues that @andersk raised, since we can just default to `None` instead of a number.

**Comment by charliermarsh** [comment]
Should have time to merge this tonight if it's in good shape!

**Comment by evanrittenhouse** [comment]
@charliermarsh just have to get the new changes re-reviewed - I changed the logic for determining successive pairs to simply be a difference of 1 between start of first/second slices. That'll allow us to pick up cases like `zip(x[2:], x[3:])` which should still trigger the violation, e.g. `pairwise(x[2:])`.

**Comment by charliermarsh** [comment]
Thanks @evanrittenhouse!
